File: momentra/momentra/Shell/GroupActive/Shared/GroupMomentsContributionComponents.swift
import SwiftUI

typealias GroupContributionItem = APIClient.GroupContributionItem

enum ContributionPaymentMethodLabel {
    static func display(_ code: String?) -> String {
        switch (code ?? "").uppercased() {
        case "UPI": return "UPI"
        case "BANK_TRANSFER": return "Bank Transfer"
        case "CASH": return "Cash"
        case "CARD": return "Card"
        default: return code?.isEmpty == false ? code! : "—"
        }
    }
}

func contributionParseDate(_ iso: String?) -> Date? {
    guard let iso, !iso.isEmpty else { return nil }
    let withFraction = ISO8601DateFormatter()
    withFraction.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    if let d = withFraction.date(from: iso) { return d }
    let plain = ISO8601DateFormatter()
    plain.formatOptions = [.withInternetDateTime]
    return plain.date(from: iso)
}

func contributionRelativeTime(_ iso: String?) -> String {
    guard let date = contributionParseDate(iso) else { return "" }
    let seconds = Int(-date.timeIntervalSinceNow)
    if seconds < 60 { return "just now" }
    if seconds < 3600 { return "\(seconds / 60)m ago" }
    if seconds < 86_400 { return "\(seconds / 3600)h ago" }
    if seconds < 172_800 { return "Yesterday" }
    let fmt = DateFormatter()
    fmt.dateFormat = "d MMM"
    return fmt.string(from: date)
}

func contributionDayKey(_ iso: String?) -> Date? {
    guard let date = contributionParseDate(iso) else { return nil }
    return Calendar.current.startOfDay(for: date)
}

func contributionDayHeader(_ day: Date, today: Date = Calendar.current.startOfDay(for: Date())) -> String {
    let cal = Calendar.current
    if cal.isDate(day, inSameDayAs: today) { return "Today" }
    if let y = cal.date(byAdding: .day, value: -1, to: today), cal.isDate(day, inSameDayAs: y) {
        return "Yesterday"
    }
    let fmt = DateFormatter()
    fmt.dateFormat = "EEEE, d MMM"
    return fmt.string(from: day)
}

private let contributionAvatarColors: [Color] = [
    Color(hex: "#FDBA74"), Color(hex: "#86EFAC"), Color(hex: "#F9A8D4"), Color(hex: "#93C5FD"),
]

struct MomentsContributionCard: View {
    let item: GroupContributionItem
    var chrome: MomentsChrome
    var avatarIndex: Int = 0
    var onTap: (() -> Void)? = nil

    private var name: String {
        let n = item.displayName?.trimmingCharacters(in: .whitespacesAndNewlines)
        return (n?.isEmpty == false) ? n! : "Member"
    }

    private var initials: String {
        let parts = name.split(separator: " ")
        if parts.count >= 2 {
            return "\(parts[0].prefix(1))\(parts[1].prefix(1))".uppercased()
        }
        return String(name.prefix(2)).uppercased()
    }

    private var meta: String {
        let time = contributionRelativeTime(item.contributedAt)
        let pool = item.label?.trimmingCharacters(in: .whitespacesAndNewlines)
        let poolPart = (pool?.isEmpty == false) ? pool! : "Pool"
        if time.isEmpty { return poolPart }
        return "\(time) · \(poolPart)"
    }

    private var amountText: String {
        GroupFinanceFormat.formatMoney(item.amount, currencyCode: item.currencyCode ?? "INR")
    }

    private var statusLabel: String {
        (item.status ?? "PAID").uppercased() == "PENDING" ? "PENDING" : "PAID"
    }

    var body: some View {
        Button {
            onTap?()
        } label: {
            VStack(alignment: .leading, spacing: 12) {
                HStack(spacing: 12) {
                    Text(initials)
                        .font(.plusJakarta(size: 14, weight: .bold))
                        .foregroundStyle(Color(hex: "#14121B"))
                        .frame(width: 40, height: 40)
                        .background(contributionAvatarColors[avatarIndex % contributionAvatarColors.count])
                        .clipShape(Circle())
                    VStack(alignment: .leading, spacing: 2) {
                        Text(name)
                            .font(.plusJakarta(size: 14, weight: .bold))
                            .foregroundStyle(chrome.text)
                        Text(meta)
                            .font(.plusJakarta(size: 11))
                            .foregroundStyle(chrome.secondary)
                    }
                    Spacer(minLength: 8)
                    if item.hasAttachment {
                        Image(systemName: "paperclip")
                            .font(.system(size: 12, weight: .semibold))
                            .foregroundStyle(chrome.accent)
                    }
                    Text(statusLabel)
                        .font(.plusJakarta(size: 10, weight: .bold))
                        .foregroundStyle(chrome.text)
                        .padding(.horizontal, 6)
                        .padding(.vertical, 2)
                        .background(chrome.accent.opacity(0.1))
                        .clipShape(RoundedRectangle(cornerRadius: 4))
                }
                HStack {
                    Text(amountText)
                        .font(.plusJakarta(size: 13, weight: .semibold))
                        .foregroundStyle(chrome.text)
                    Spacer()
                    Text(ContributionPaymentMethodLabel.display(item.paymentMethodCode))
                        .font(.plusJakarta(size: 11))
                        .foregroundStyle(chrome.secondary)
                }
            }
            .padding(14)
            .background(chrome.card)
            .clipShape(RoundedRectangle(cornerRadius: 14))
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .disabled(onTap == nil)
    }
}

struct MomentsContributionDetailsSection: View {
    let items: [GroupContributionItem]
    var chrome: MomentsChrome
    var momentId: String? = nil
    var onViewAll: (() -> Void)?
    var onEdit: ((GroupContributionItem) -> Void)? = nil

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            MomentsSectionHeader(title: "Contribution Details", chrome: chrome, onViewAll: onViewAll)
            if items.isEmpty {
                GroupEmptySection(
                    message: "No contributions yet",
                    detail: "Record a contribution from Quick Add — nothing is invented."
                )
            } else {
                ForEach(Array(items.prefix(3).enumerated()), id: \.element.id) { idx, item in
                    MomentsContributionCard(
                        item: item,
                        chrome: chrome,
                        avatarIndex: idx,
                        onTap: (momentId != nil && onEdit != nil) ? { onEdit?(item) } : nil
                    )
                }
            }
        }
    }
}

struct ContributionsListSheet: View {
    let items: [GroupContributionItem]
    var chrome: MomentsChrome
    var momentId: String? = nil
    var onDismiss: () -> Void
    var onEdit: ((GroupContributionItem) -> Void)? = nil

    private var grouped: [(day: Date, items: [GroupContributionItem])] {
        let today = Calendar.current.startOfDay(for: Date())
        var buckets: [Date: [GroupContributionItem]] = [:]
        var order: [Date] = []
        for item in items {
            let day = contributionDayKey(item.contributedAt) ?? today
            if buckets[day] == nil {
                order.append(day)
                buckets[day] = []
            }
            buckets[day]?.append(item)
        }
        return order.sorted(by: >).compactMap { day in
            guard let list = buckets[day], !list.isEmpty else { return nil }
            return (day, list)
        }
    }

    var body: some View {
        NavigationStack {
            ScrollView {
                LazyVStack(alignment: .leading, spacing: 16) {
                    if grouped.isEmpty {
                        GroupEmptySection(
                            message: "No contributions yet",
                            detail: "Record a contribution from Quick Add — nothing is invented."
                        )
                        .padding(.top, 24)
                    } else {
                        ForEach(Array(grouped.enumerated()), id: \.offset) { _, group in
                            VStack(alignment: .leading, spacing: 10) {
                                Text(contributionDayHeader(group.day))
                                    .font(.plusJakarta(size: 13, weight: .bold))
                                    .foregroundStyle(chrome.secondary)
                                ForEach(Array(group.items.enumerated()), id: \.element.id) { idx, item in
                                    MomentsContributionCard(
                                        item: item,
                                        chrome: chrome,
                                        avatarIndex: idx,
                                        onTap: (momentId != nil && onEdit != nil) ? { onEdit?(item) } : nil
                                    )
                                }
                            }
                        }
                    }
                }
                .padding(20)
            }
            .background(chrome.bg)
            .navigationTitle("Contributions")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Close", action: onDismiss)
                        .foregroundStyle(chrome.accent)
                }
            }
        }
        .presentationDetents([.medium, .large])
    }
}

File: apk/app/src/main/java/com/example/momentra/ui/shell/group/shared/GroupMomentsContributionComponents.kt
package com.example.momentra.ui.shell.group.shared

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.momentra.data.api.GroupContributionItemDto
import com.example.momentra.ui.theme.PlusJakartaSans
import java.time.Instant
import java.time.LocalDate
import java.time.OffsetDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import java.util.Locale

private val ContributionAvatarColors = listOf(
    Color(0xFFFDBA74), Color(0xFF86EFAC), Color(0xFFF9A8D4), Color(0xFF93C5FD),
)

fun contributionPaymentMethodLabel(code: String?): String = when (code?.trim()?.uppercase(Locale.US)) {
    "UPI" -> "UPI"
    "BANK_TRANSFER" -> "Bank Transfer"
    "CASH" -> "Cash"
    "CARD" -> "Card"
    else -> code?.takeIf { it.isNotBlank() } ?: "—"
}

private fun parseContributionInstant(iso: String?): Instant? {
    if (iso.isNullOrBlank()) return null
    return runCatching { Instant.parse(iso) }.getOrNull()
        ?: runCatching { OffsetDateTime.parse(iso).toInstant() }.getOrNull()
}

fun contributionRelativeTime(iso: String?): String {
    val instant = parseContributionInstant(iso) ?: return ""
    val seconds = ChronoUnit.SECONDS.between(instant, Instant.now()).toInt()
    return when {
        seconds < 60 -> "just now"
        seconds < 3600 -> "${seconds / 60}m ago"
        seconds < 86_400 -> "${seconds / 3600}h ago"
        seconds < 172_800 -> "Yesterday"
        else -> DateTimeFormatter.ofPattern("d MMM", Locale.US)
            .format(instant.atZone(ZoneId.systemDefault()))
    }
}

fun contributionDayKey(iso: String?): LocalDate? {
    val instant = parseContributionInstant(iso) ?: return null
    return instant.atZone(ZoneId.systemDefault()).toLocalDate()
}

fun contributionDayHeader(day: LocalDate, today: LocalDate = LocalDate.now()): String = when (day) {
    today -> "Today"
    today.minusDays(1) -> "Yesterday"
    else -> DateTimeFormatter.ofPattern("EEEE, d MMM", Locale.US).format(day)
}

private fun initials(name: String): String {
    val parts = name.trim().split(Regex("\\s+")).filter { it.isNotEmpty() }
    return if (parts.size >= 2) {
        "${parts[0].first()}${parts[1].first()}".uppercase(Locale.US)
    } else {
        name.take(2).uppercase(Locale.US)
    }
}

private fun amountLabel(amount: String?, currency: String?): String {
    val code = currency?.uppercase(Locale.US) ?: "INR"
    val symbol = when (code) {
        "INR" -> "₹"
        "USD" -> "$"
        "EUR" -> "€"
        "GBP" -> "£"
        else -> "$code "
    }
    return "$symbol${amount ?: "0"}"
}

@Composable
fun MomentsContributionCard(
    item: GroupContributionItemDto,
    chrome: MomentsChrome,
    avatarIndex: Int = 0,
    onTap: (() -> Unit)? = null,
) {
    val name = item.displayName?.takeIf { it.isNotBlank() } ?: "Member"
    val pool = item.label?.takeIf { it.isNotBlank() } ?: "Pool"
    val time = contributionRelativeTime(item.contributedAt)
    val meta = if (time.isBlank()) pool else "$time · $pool"
    val status = if (item.status.equals("PENDING", true)) "PENDING" else "PAID"

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(16.dp))
            .background(chrome.card)
            .border(1.dp, chrome.border, RoundedCornerShape(16.dp))
            .then(
                if (onTap != null) Modifier.clickable(onClick = onTap) else Modifier,
            ),
    ) {
        Row {
            Box(
                modifier = Modifier
                    .width(4.dp)
                    .height(104.dp)
                    .background(chrome.accent),
            )
            Column(
                modifier = Modifier
                    .weight(1f)
                    .padding(16.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                Row(
                    horizontalArrangement = Arrangement.spacedBy(12.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Box(
                        modifier = Modifier
                            .size(40.dp)
                            .clip(CircleShape)
                            .background(ContributionAvatarColors[avatarIndex % ContributionAvatarColors.size]),
                        contentAlignment = Alignment.Center,
                    ) {
                        Text(
                            initials(name),
                            color = Color(0xFF14121B),
                            fontSize = 14.sp,
                            fontWeight = FontWeight.Bold,
                            fontFamily = PlusJakartaSans,
                        )
                    }
                    Column(modifier = Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
                        Text(name, color = chrome.text, fontSize = 14.sp, fontWeight = FontWeight.Bold, fontFamily = PlusJakartaSans)
                        Text(meta, color = chrome.secondary, fontSize = 11.sp, fontFamily = PlusJakartaSans)
                    }
                    if (item.hasAttachment) {
                        AttachmentPaperclipIcon(tint = chrome.accent)
                    }
                    Text(
                        status,
                        color = chrome.text,
                        fontSize = 10.sp,
                        fontWeight = FontWeight.Bold,
                        fontFamily = PlusJakartaSans,
                        modifier = Modifier
                            .clip(RoundedCornerShape(4.dp))
                            .background(chrome.accent.copy(alpha = 0.1f))
                            .padding(horizontal = 6.dp, vertical = 2.dp),
                    )
                }
                Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
                    Text(
                        amountLabel(item.amount, item.currencyCode),
                        color = chrome.text,
                        fontSize = 13.sp,
                        fontWeight = FontWeight.SemiBold,
                        fontFamily = PlusJakartaSans,
                    )
                    Text(
                        contributionPaymentMethodLabel(item.paymentMethodCode),
                        color = chrome.secondary,
                        fontSize = 11.sp,
                        fontFamily = PlusJakartaSans,
                    )
                }
            }
        }
    }
}

@Composable
fun MomentsContributionDetailsSection(
    items: List<GroupContributionItemDto>,
    chrome: MomentsChrome,
    momentId: String? = null,
    onViewAll: (() -> Unit)? = null,
    onEdit: ((GroupContributionItemDto) -> Unit)? = null,
) {
    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        MomentsSectionHeader(title = "Contribution Details", chrome = chrome, onViewAll = onViewAll)
        if (items.isEmpty()) {
            GroupEmptySection(
                message = "No contributions yet",
                detail = "Record a contribution from Quick Add — nothing is invented.",
            )
        } else {
            items.take(3).forEachIndexed { idx, item ->
                MomentsContributionCard(
                    item = item,
                    chrome = chrome,
                    avatarIndex = idx,
                    onTap = if (momentId != null && onEdit != null) {
                        { onEdit(item) }
                    } else {
                        null
                    },
                )
            }
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ContributionsListSheet(
    items: List<GroupContributionItemDto>,
    visible: Boolean,
    onDismiss: () -> Unit,
    chrome: MomentsChrome,
    momentId: String? = null,
    onEdit: ((GroupContributionItemDto) -> Unit)? = null,
) {
    if (!visible) return
    val grouped = remember(items) {
        val today = LocalDate.now()
        val map = linkedMapOf<LocalDate, MutableList<GroupContributionItemDto>>()
        items.forEach { item ->
            val day = contributionDayKey(item.contributedAt) ?: today
            map.getOrPut(day) { mutableListOf() }.add(item)
        }
        map.entries.sortedByDescending { it.key }
    }
    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true),
        containerColor = chrome.bg,
        dragHandle = {
            Box(
                modifier = Modifier
                    .padding(top = 12.dp, bottom = 4.dp)
                    .size(width = 40.dp, height = 5.dp)
                    .clip(RoundedCornerShape(100.dp))
                    .background(Color.White.copy(alpha = 0.2f)),
            )
        },
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .navigationBarsPadding()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 20.dp)
                .padding(bottom = 28.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp),
        ) {
            Text(
                "Contributions",
                color = chrome.text,
                fontSize = 18.sp,
                fontWeight = FontWeight.Bold,
                fontFamily = PlusJakartaSans,
            )
            if (grouped.isEmpty()) {
                GroupEmptySection(
                    message = "No contributions yet",
                    detail = "Record a contribution from Quick Add — nothing is invented.",
                )
            } else {
                grouped.forEach { (day, dayItems) ->
                    Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
                        Text(
                            contributionDayHeader(day),
                            color = chrome.secondary,
                            fontSize = 13.sp,
                            fontWeight = FontWeight.Bold,
                            fontFamily = PlusJakartaSans,
                        )
                        dayItems.forEachIndexed { idx, item ->
                            MomentsContributionCard(
                                item = item,
                                chrome = chrome,
                                avatarIndex = idx,
                                onTap = if (momentId != null && onEdit != null) {
                                    { onEdit(item) }
                                } else {
                                    null
                                },
                            )
                        }
                    }
                }
            }
        }
    }
}
